snapshots/devmapper/metadata.go

```diff
@@ -122,11 +121,15 @@ func (m *PoolMetadata) AddDevice(ctx context.Context, info *DeviceInfo) error {
 	return nil
 }
 
-// MarkFaulty marks the given device and corresponding devmapper device ID as faulty.
+// MarkFaulty marks the given device as faulty.
+//
+// TODO (need clarfication): how does this work? With `another devmapper ID`, it should
+// refers to another device, so it's not clear how it refers to `Faulty` state and recreate
+// it.
```

Comment:
In addition to device metadata table, we keep device ids states table, in order to track which id is used and which to consume next (so `getNextDeviceID` returns next `deviceFree` ID).
Sometimes devmapper can't create a device with given ID and snapshot creation fails (for whatever reason). In this case we mark this ID as `deviceFaulty` in order to prevent `getNextDeviceID` from returning this ID again for the subsequent layer calls (otherwise the devmapper shapshotter will entirely stuck on the first wrong ID). 
So this way the snapshotter can move forward with other ID and leave this one for manual check/resolve/fix.

Reply:
Thanks.

> In addition to device metadata table, we keep device ids states table, in order to track which id is used and which to consume next (so `getNextDeviceID` returns next `deviceFree` ID).

Yes, I see, but deviceTaken/deviceFree seems enough and straight for the purpose.

> Sometimes devmapper can't create a device with given ID and snapshot creation fails (for whatever reason). In this case we mark this ID as `deviceFaulty` in order to prevent 

Mark the deviceState of device info in device table as `Faulty`, I think, can avoid upper layer to retry with the same SnapID ($thinpool-snap-$SnapID), while deviceID table is an internal to metadata module, should not directly have something with upper layer logical.

`getNextDeviceID` from returning this ID again for the subsequent layer calls (otherwise the devmapper shapshotter will entirely stuck on the first wrong ID).

Mark the deviceID `Taken` looks enough?

> So this way the snapshotter can move forward with other ID and leave this one for manual check/resolve/fix.

I feel the object we should directly refer to is device/snapshot name:-)

Reply:
Can you elaborate what kind of problem you're trying to solve by these changes? I might be loosing some context here, but I don't understand the benefits.

> while deviceID table is an internal to metadata module, should not directly have something with upper layer logical.

ID is the only way to refer to a devmapper device or snapshot. I'd prefer to keep this reference and track their states properly.

> Mark the deviceState of device info in device table as Faulty, I think, can avoid upper layer to retry with the same SnapID ($thinpool-snap-$SnapID),

Retry may success for the same snapshot ID (with another devmapper ID). This way it is possible to troubleshoot failed ID (by tracking it in the device ID table), while allowing the snapshotter to move forward with other layers.

> Mark the deviceID Taken looks enough?

Marking failed devmapper ID as `Taken` doesn't sound right to me.

Reply:
> Can you elaborate what kind of problem you're trying to solve by these changes? I might be loosing some context here, but I don't understand the benefits.
> 

The first patch to remove Faulty deviceID state is not required to solve this issue. When looking at the code, I think  it's useless because we already have Faulty state for deviceInfo.

Alright, it's my fault to bring in a unrelated change here, will drop it :-)

> > Mark the deviceID Taken looks enough?
> 
> Marking failed devmapper ID as `Taken` doesn't sound right to me.

While I think it's duplicate state with faulty device state, but yeah on the other hand, it's also clear flag within deviceID table. OK, it's keep it.

Reply:
Sorry, it might be confusing. To sum up, the difference is:
- Faulty devmapper ID - we failed and it needs manual intervention by user.
- Faulty device - failed for the given devmapper ID, but can be retried by snapshotter next time (and potentially succeed).

Reply:
> * Faulty devmapper ID - we failed and it needs manual intervention by user.

Ah, I'd like to call it thin device ID or just device ID. I interpreted `devmapper ID` as snapID which suffixes in $pool-snap-$snapID.